src/InputAreaWidget.ts

```diff
@@ -0,0 +1,270 @@
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  ResizeMessage, Widget, Panel
+} from 'phosphor-widget';
+
+import {
+  IInputAreaViewModel, ITextEditorViewModel
+} from './InputAreaViewModel';
+
+
+let diffMatchPatch = new dmp.diff_match_patch()
+
+
+/**
+ * A widget which hosts a CodeMirror editor.
+ */
+export
+class CodeMirrorWidget extends Widget {
+
+  /**
+   * Construct a CodeMirror widget.
+   */
+  constructor(model: ITextEditorViewModel) {
+    super();
+    this.addClass('jp-CodeMirrorWidget');
+    this._model = model;
+    this._editor = CodeMirror(this.node);
+    this.updateMimetype(model.mimetype);
+    this.updateLineNumbers(model.lineNumbers);
+    this.updateFixedHeight(model.fixedHeight);
+    this.updateText(model.text);
+
+    this._editor.on('change', (instance, change) => {
+      this._model.text = this._editor.getDoc().getValue();
+    });
+    model.stateChanged.connect(this._modelUpdate, this);
+  }
+
+  /**
+   * Update whether the editor has a fixed maximum height.
+   */
+  updateFixedHeight(fixedHeight: boolean) {
+    if (fixedHeight) {
+      this.addClass('jp-CodeMirrorWidget-fixedHeight');
+    } else {
+      this.removeClass('jp-CodeMirrorWidget-fixedHeight');
+    }
+  }
+
+  /**
+   * Update the text in the widget.
+   *
+   * #### Notes
+   * This function attempts to restore the cursor to the correct
+   * place by using the bitap algorithm to find the corresponding
+   * position of the cursor in the new text.
+   */
+  updateText(text: string) {
+    if (this.text !== text) {
+      // TODO: do something smart with all the selections
+
+      let doc = this._editor.getDoc();
+      let oldText = doc.getValue();
+      let oldCursor = doc.indexFromPos(doc.getCursor());
+      let cursor = 0;
+      if (oldCursor === oldText.length) {
+        // if the cursor was at the end, keep it at the end
+        cursor = text.length;
+      } else {
+        let fragment = oldText.substr(oldCursor, 10);
+        cursor = diffMatchPatch.match_main(text, fragment, oldCursor);
+      }
+
+      doc.setValue(text);
+      doc.setCursor(doc.posFromIndex(cursor));
+    }
+  }
+
+  /**
+   * Set the mode by giving the mimetype.
+   *
+   * #### Notes
+   * Valid mimetypes are listed in https://github.com/codemirror/CodeMirror/blob/master/mode/meta.js.
+   */
+  updateMimetype(mimetype: string) {
+    if (CodeMirror.mimeModes.hasOwnProperty(mimetype)) {
+      this._editor.setOption('mode', mimetype);
+    } else {
+      let info = CodeMirror.findModeByMIME(mimetype);
+      if (info) {
+        this._loadCodeMirrorMode(info.mode).then(() => {
+          this._editor.setOption('mode', mimetype);
+        })
+      }
+    }
+  }
+
+  /**
+   * Update the line numbers in the editing widget.
+   */
+  updateLineNumbers(lineNumbers: boolean) {
+    this._editor.setOption('lineNumbers', lineNumbers);
+  }
+
+  /**
+   * Get the text in the widget.
+   */
+  get text(): string {
```

Comment:
I don't want to expose CodeMirror (unless there is a good usecase for making code specific to codemirror), so I don't want to expose the `_editor`.  Do you mean a read-only getter for the editor model?